Upgrade gradle to 7.1.1 agp to 4.2.2

oguzkocer opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

This PR upgrades react-native Android libraries to Gradle 7 and Android Gradle Plugin 4.2.2. Most of the changes are generated by the Gradle upgrade script: `./gradlew wrapper --gradle-version=7.1.1 --distribution-type=all`. It also includes the following changes:

* Updates several react-native library versions to new tags which contains Gradle 7 upgrades
* Utilizes plugin DSL for Gradle to simplify how Gradle plugins are set across projects. This will help us avoid upgrading the react-native libraries we use in the future just so we can use the latest Gradle plugin (unless the upgrade requires changes in the project)

Related gb-mobile PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3833

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

Manual testing of the react-native demo app and [WordPress-Android PR](https://github.com/wordpress-mobile/WordPress-Android/pull/15076) that includes these changes.

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Android build file changes for react-native projects & react-native library updates for Android projects.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

oguzkocer committed:
Upgrade Android projects to Gradle 7.1.1

oguzkocer committed:
Upgrade Android projects to AGP 4.2.2

oguzkocer committed:
Remove unnecessary maven plugin from react-native-aztec

oguzkocer committed:
Temporarily use Jitpack for all node_modules project dependencies

oguzkocer committed:
Prefer android:color/white over color/white in react-native-bridge

oguzkocer committed:
Update react-native-bridge project dependencies from implementation to api

oguzkocer committed:
Update react-native-svg & react-native-slider

oguzkocer committed:
Use plugin DSL in react-native-bridge

oguzkocer committed:
Re-add react-native-slider and react-native-svg to react-native-bridge android

oguzkocer committed:
Update react-native-linear-gradient & react-native-video

oguzkocer committed:
Updates remaining react-native libraries to Gradle 7

oguzkocer committed:
Revert react-native-bridge project dependency changes

oguzkocer committed:
Re-add accidentally removed react-native-webview and update its version

oguzkocer committed:
Update react-native-aztec to use Plugin DSL

oguzkocer committed:
Adds agp & kotlin plugins (without applying) to root react-native-bridge build.gradle

oguzkocer committed:
Remove explicit kotlin dependency from react-native-bridge

oguzkocer committed:
Update react-native libraries: masked-view, reanimated, safe-area-context, screens

oguzkocer committed:
Merge remote-tracking branch 'origin/trunk' into upgrade-gradle-to-7.1.1-agp-to-4.2.2

oguzkocer committed:
Update react-native-hsv-color-picker

illusaen committed:
Updating package.json

Updating package.json for libraries to use new wp-fork for Gradle 7

illusaen committed:
Updating package.json

Including react-native-prompt-android wp-fork change.

mchowning commented on packages/react-native-editor/package.json:
```diff
@@ -31,8 +31,8 @@
 		"@react-native-community/blur": "3.6.0",
-		"@react-native-community/masked-view": "git+https://github.com/wordpress-mobile/react-native-masked-view.git#v0.1.11-wp",
-		"@react-native-community/slider": "git+https://github.com/wordpress-mobile/react-native-slider.git#v3.0.2-wp",
+		"@react-native-community/masked-view": "git+https://github.com/wordpress-mobile/react-native-masked-view.git#wp-fork",
```
I think we need to not use `wp-fork` for any of these versions because even though there is a `wp-fork` tag, it conflicts with the `wp-fork` branch on that repo and in my testing, npm will pull the branch, not the tag, when they conflict. 

[Related comment](https://github.com/wordpress-mobile/react-native-hsv-color-picker/pull/3#issuecomment-902950210)

oguzkocer replied:
@mchowning @illusaen Shouldn't it be `v0.1.12-wp`, `v0.2.11-wp` or something similar, because the `-wp` ones are our tags, right?

I think @ceyhun might have been the one who worked on the previous tags for react-native `0.64` upgrade, maybe he can advise on how to proceed?

This is one of the reasons I left the tagging/updating the references part to gutenberg-mobile developers, because I thought there was a process in place. And if there isn't one, maybe we can settle on one and document it now?

mchowning replied:
In general, I believe that the `v###-wp` signifies that it is `v###` of the library, and[ we add the `-wp` to signify it is from our fork](https://github.com/wordpress-mobile/react-native-masked-view/commit/959000abf16d7a024c9c084af7a9fc7d86293573). The issue I was thinking we would have with that is that if we didn't update the library itself to a new version (i.e., only added our gradle updates), we couldn't follow that same scheme because the library version is unchanged so it would be the same tag (I guess technically we could "move" the tag, but that's a bad idea for a number of reasons). ~With that said, as you point out `masked-view` is a new version for some reason (and it wasn't updated through a PR, so we probably need to lock down those `wp-fork` branches to only be updated by PRs), which I wasn't expecting.~ (EDIT: I was looking at the `master` branch instead of `wp-fork` by mistake)

ceyhun replied:
> This is one of the reasons I left the tagging/updating the references part to gutenberg-mobile developers, because I thought there was a process in place. And if there isn't one, maybe we can settle on one and document it now?

Just to close the loop here, I think we've settled on tagging with `[upstream version]-wp-N`, like `v11.6.5-wp-1`

illusaen committed:
Update package.json

mchowning committed:
Bumping forked dependency versions for Gradle 7

mchowning committed:
Update package-lock file

By running: `npm uninstall @wordpress/react-native-editor --save && npm
install ./packages/react-native-editor --save`

mchowning committed:
Merge remote-tracking branch 'origin/trunk' into upgrade-gradle-to-7.1.1-agp-to-4.2.2

mchowning committed:
Merge remote-tracking branch 'origin/trunk' into upgrade-gradle-to-7.1.1-agp-to-4.2.2

Commit:
Update react-native-gesture-handler fork

Bring in the latest changes from react-native-gesture-handler's wp-fork branch

Commit:
Temp force new NPM cache

Commit:
Bump react-native-gesture-handler

Commit:
Bump react-native-gesture-handler again

Commit:
Bump react-native-gesture-handler

Commit:
Temp force new NPM cache

Commit:
Change preactjs/compressed-size-action to npm i

Commit:
Change preactjs/compressed-size-action to npm install

ceyhun committed:
Revert "Change preactjs/compressed-size-action to npm install"

This reverts commit 6073a6b24aa0c3266e2416f0a9af3907fb4dcc79.

ceyhun committed:
Revert "Change preactjs/compressed-size-action to npm i"

This reverts commit bba53beaa14c8ec179ce7a62022bd040e1b89e20.

ceyhun committed:
Revert "Temp force new NPM cache"

This reverts commit 0260fa4bee59fde7e5b0de8ec345798cdbfbbfa7.

ceyhun committed:
Revert "Temp force new NPM cache"

This reverts commit 1d9bdb3b12c413320373c92342f08fa41f2c55f5.

mchowning committed:
Merge remote-tracking branch 'origin/trunk' into upgrade-gradle-to-7.1.1-agp-to-4.2.2

mchowning committed:
Update react-native-gesture-handler version

mchowning committed:
Force CI cache to update (must be reverted before merge)

mchowning committed:
Revert "Force CI cache to update (must be reverted before merge)"

This reverts commit ecb1829de5990058796bd91ace167b6a53ec87fa.

mchowning reviewed (approved):
Looks good! Thanks for this @oguzkocer !

mchowning reviewed (changes requested):
I forgot to test WPAndroid before approving this, and I am seeing the composite build failing there, so I'll remove my approval to make sure we figure out what is going on there. 

This is the error message I am seeing.

```
code/a8c/WordPress-Android  c1733ec486 ✔                                                                                                                                                29d21h
▶ ./gradlew installWordPressWasabiDebug
Configuration on demand is an incubating feature.
Substituting gutenberg-mobile with the local build

> Configure project :libs:stories-android
Checking whether `configure` binary is present

> Configure project :libs:stories-android:stories
WARNING: DSL element 'android.viewBinding.enabled' is obsolete and has been replaced with 'android.buildFeatures.viewBinding'.
It will be removed in version 5.0 of the Android Gradle plugin.

FAILURE: Build failed with an exception.

* Where:
Build file '/Users/matt/code/a8c/gutenberg-mobile/gutenberg/node_modules/@react-native-community/masked-view/android/build.gradle' line: 2

* What went wrong:
An exception occurred applying plugin request [id: 'com.android.library', version: '4.2.2']
> Failed to apply plugin [id 'com.android.internal.version-check']
   > Minimum supported Gradle version is 6.7.1. Current version is 6.1.1. If using the gradle wrapper, try editing the distributionUrl in /Users/matt/code/a8c/gutenberg-mobile/gutenberg/packages/react-native-bridge/android/gradle/wrapper/gradle-wrapper.properties to gradle-6.7.1-all.zip

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 5s

code/a8c/WordPress-Android  c1733ec486 ✔                                                                                                                                               29d21h  ⍉
▶ cat  /Users/matt/code/a8c/gutenberg-mobile/gutenberg/packages/react-native-bridge/android/gradle/wrapper/gradle-wrapper.properties
distributionBase=GRADLE_USER_HOME
distributionPath=wrapper/dists
distributionUrl=https\://services.gradle.org/distributions/gradle-7.1.1-all.zip
zipStoreBase=GRADLE_USER_HOME
zipStorePath=wrapper/dists
```

oguzkocer commented:
As mentioned on Slack, the build failure is due to testing the build on a non Gradle-7 branch. @mchowning is doing another set of tests and will re-approve this PR soon. 🤞

mchowning reviewed (approved):
Sorry about my mix-up with the WPAndroid branch. Tested again with the proper branch and, shockingly, things worked so much better! 🤦  Re-approving now.

jeffpaul commented:
@illusaen pinging you here to see if you can confirm your WordPress.org username so that we can properly credit you in the WordPress 5.9 release?  Thanks!
